### PR TITLE
Breaking changes for FileStream

### LINKED ISSUE #23857: Breaking change: FileStream does not synchronize file offset with OS anymore
##  FileStream does not synchronize file offset with OS anymore

@adamsitnik described it in detail here: https://github.com/dotnet/runtime/issues/50860

cc @jozkee

### Version introduced

.NET 6.0 Preview 4

### Recommended action

We will write a blog post with recommended practices.

### Category

- [X] Core .NET libraries

### Affected APIs

System.IO.FileStream APIs that used to update the OS file offset.

Finally, please remember to email a link to this breaking change issue to [.NET Breaking Change Notifications](mailto:dotnetbcn@microsoft.com).

---
#### Issue metadata

* Issue type: breaking-change

### LINKED ISSUE #23858: Breaking change: FileStream.Position is updated AFTER ReadAsync|WriteAsync completes
## FileStream.Position is updated AFTER ReadAsync|WriteAsync completes

@adamsitnik described it in detail here: https://github.com/dotnet/runtime/issues/50858

cc @jozkee

### Version introduced

.NET 6.0 Preview 4

### Recommended action

We will write a blog post and will share good practices.

### Category

- [C] Core .NET libraries

### Affected APIs

FileStream.ReadAsync and FileStream.WriteAsync.

#### Issue metadata

* Issue type: breaking-change

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #23857 
Fixes #23858 

- [File offset preview](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/filestream-doesnt-sync-offset-with-os?branch=pr-en-us-24060)
- [Position preview](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/filestream-position-updates-after-readasync-writeasync-completion?branch=pr-en-us-24060)

### CODE REVIEW

File: docs/core/compatibility/core-libraries/6.0/filestream-doesnt-sync-offset-with-os.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+This change was introduced to improve the performance of asynchronous reads and writes and to address the following issues:
+
+- [Win32 FileStream will issue a seek on every ReadAsync call](https://github.com/dotnet/runtime/issue/16354)
+- [FileStream.Windows useAsync WriteAsync calls blocking APIs](https://github.com/dotnet/runtime/issue/25905)
+
+With this change, <xref:System.IO.FileStream.ReadAsync%2A> operations are up to two times faster, and <xref:System.IO.FileStream.WriteAsync%2A> operations are up to five times faster.
+
+## Recommended action
+
+
```
Comment (gewarren) [inline docs/core/compatibility/core-libraries/6.0/filestream-doesnt-sync-offset-with-os.md:52]: Please help me fill this out with a suggestion :)

File: docs/core/compatibility/core-libraries/6.0/filestream-position-updates-after-readasync-writeasync-completion.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+This change was introduced to allow for 100% asynchronous file I/O with <xref:System.IO.FileStream> and to fix the following issues:
+
+- [FileStream.FlushAsync ends up doing synchronous writes](https://github.com/dotnet/runtime/issue/27643)
+- [Win32 FileStream turns async reads into sync reads](https://github.com/dotnet/runtime/issue/16341)
+
+Now, when buffering is enabled (that is, the `bufferSize` argument that's passed to the [FileStream constructor](xref:System.IO.FileStream.%23ctor%2A) is greater than 1), every <xref:System.IO.FileStream.ReadAsync%2A> and <xref:System.IO.FileStream.WriteAsync%2A> operation is serialized.
+
+## Recommended action
+
+
```
Comment (gewarren) [inline docs/core/compatibility/core-libraries/6.0/filestream-position-updates-after-readasync-writeasync-completion.md:55]: Please help me fill this out with a suggestion :)

File: docs/core/compatibility/core-libraries/6.0/filestream-doesnt-sync-offset-with-os.md (hunk 1 above)
Comment (jeffhandley) [reply docs/core/compatibility/core-libraries/6.0/filestream-doesnt-sync-offset-with-os.md:52]: @dotnet/area-system-io How does this sound?

Modify any code that relied on the offset being synchronized. To enable the .NET 5 behavior in .NET 6, specify an `AppContext` switch or an environment variable. This switch is expected to be removed in .NET 7, and using it opts out of all performance improvements made to FileStream in .NET 6.

```xml
{
    "configProperties": {
        "System.IO.UseNet5CompatFileStream": true
    }
}

```

```cmd
set DOTNET_SYSTEM_IO_USENET5COMPATFILESTREAM=1
```

File: docs/core/compatibility/core-libraries/6.0/filestream-position-updates-after-readasync-writeasync-completion.md (hunk 2 above)
Comment (jeffhandley) [reply docs/core/compatibility/core-libraries/6.0/filestream-position-updates-after-readasync-writeasync-completion.md:55]: @dotnet/area-system-io How does this sound?

Modify any code that relied on the position being set before operations completed. To enable the .NET 5 behavior in .NET 6, specify an `AppContext` switch or an environment variable. This switch is expected to be removed in .NET 7, and using it opts out of all performance improvements made to FileStream in .NET 6.

```xml
{
    "configProperties": {
        "System.IO.UseNet5CompatFileStream": true
    }
}

```

```cmd
set DOTNET_SYSTEM_IO_USENET5COMPATFILESTREAM=1
```

**Review by tdykstra — APPROVED** [review]
LGTM

**Comment by gewarren** [comment]
Thank you for reviewing it Tom!

**Pull request merged**